3단계 - 질문 삭제하기 리팩터링 PR 요청드립니다.

earth-h opened the pull request:
안녕하세요, 민수님 :) 

3단계 - 질문 삭제하기 리팩터링 PR 피드백 요청드립니다.

먼저, 리팩터링 시 중점적으로 생각했던 부분은 아래와 같습니다 :) 
- 객체에 메시지를 보내어 로직 수행하도록 진행
- Answers 일급컬렉션 생성 -> 임베디드 타입 지정
  - 일급컬렉션 사용 시, list 내부가 외부요인에 의해 변경되지 않도록 관리
    - 생성자를 통해 list를 넣을 때 복사해서 넣도록 하여 외부에 의해 의도치 않은 변경 막고자 함
    - getAnswers()와 같은 list를 반환하는 메소드 대신, list를 조회해서 하려고 하는 행위(예: size() 반환 등) 자체를 메소드로 뽑아 사용
- DeleteHistories 일급컬렉션 생성
  - Answers와 달리, DeleteHistories는 list가 조회되어야 레포지토리에 넣을 수 있어, unmodifiedList()로 반환하여 저장할 수 있도록 함
  - 생성자를 통해 list 넣을 때 복사하여 넣도록 하여 외부에 의한 의도치 않은 변경 막고자 함
- 양방향 연관관계인 Answer와 Question 사이에 연관관계 편의 메소드는 Answer에만 생성
  - 기본적으로, 질문이 생성된 후 답변이 생성되어 질문과 연결되기 때문에 Answer에 의해서만 toQuestion() 통해서 연관관계가 맺어지도록 함

리팩토링 이후에도 고민되는 포인트는 아래와 같습니다...! 
- Answers에 생성한, removeAnswer()와 addAnswer()가 question에서 호출하기 때문에 public으로 열려있는데, 이렇게 열어도 될지 고민입니다. 외부 요인에 의해 잘못된 add나 remove를 막으려면 private이어야 할 것 같은데, 이를 내부적으로 수행할 방안을 찾지 못하였습니다..😭
- Answer 각각의 validation 체크를 Answers에서도 진행한 부분
  - Answer에서만 validateSameUser() 체크를 진행하고 싶었으나, 각각 Answer의 validate 후 바로 삭제여부를 변경하게 되면, 아래와 같은 상황일 때 잘못된 결과 발생하여서 중복된 validation체크가 된 것 같아 고민이 됩니다.
  - ```text
    상황) 질문A에 대한 답변1, 답변2가 존재할 때 답변2의 작성자가 질문과 상이하다.
    이때, answer 내에서만 validateSameUser() 체크 시, 답변1은 validation 통과 후 삭제여부가 true가 되고, 이후 답변2는 validation에서 걸려 exception이 터진다. 
    => 답변1의 삭제여부가 다시 false로 변경되지 않는 문제 발생
    ``` 

매번 좋은 피드백 감사드립니다 ! ㅎㅎ

earth-h committed:
test(domain): cascadeType.REMOVE 추가 테스트

cascade REMOVE 추가 테스트 진행
상황1) Answer entity 내 Question 필드에 대해 cascadeType.REMOVE 추가
- Question에 대한 Answer1, Answer2가 모두 save여서 영속상태일 때, Answer1만 삭제하면 Question은 삭제되나 Answer2는 삭제되지 않음

상황2) Answer entity 내 Question 필드에 대해 cascadeType.REMOVE 추가 및 Question entity 내 List<Answer> 필드에 대해 cascadeType.REMOVE 추가
- Question에 대한 Answer1, Answer2가 모두 save여서 영속상태일 때,Answer1 삭제 시 Question과 Question에 연관된 Answer2 역시 삭제됨

earth-h committed:
docs(guide): README.md 업데이트

JPA : 3단계 - 질문 삭제하기 리팩터링
- 기능/프로그래밍 요구 사항 정리
- 기능 목록 정리

earth-h committed:
feat(domain): 질문 삭제 권한 확인 메소드 추가

Question
- 질문 삭제 권한 확인 메소드(validateSameUser) 생성
  - 해당 메소드에서 수행하는 작업이 동일유저인지 확인하는 작업이므로 validateSameUser로 명명

ErrorCode
- 에러코드를 한 곳에서 관리하고자 enum 생성

CannotDeleteException
- Question에서 사용하려고 하니, Unhandled exception type Exception 발생하여서 extends Exception을 RuntimeException으로 수정

earth-h committed:
feat(domain): 질문 삭제여부 변경 메소드명 수정

Question
- setDeleted() 대신 changeDeleted로 삭제여부 변경 메소드명 리팩토링

QnaService, QuestionRepositoryTest
- setDeleted() 메소드명 사용 부분 메소드명 수정

earth-h committed:
refactor(domain): QuestionTest 내 ErrorCode 사용하도록 수정

earth-h committed:
feat(domain): 답변 삭제 권한 확인 메소드 추가

Answer
- 답변 삭제 권한 확인 메소드(validateSameUser) 생성
- 해당 메소드에서 수행하는 작업이 동일유저인지 확인하는 작업이므로 validateSameUser로 명명

earth-h committed:
feat(domain): 답변 삭제여부 변경 메소드명 수정

Answer
- setDeleted() 대신 changeDeleted로 삭제여부 변경 메소드명 리팩토링

earth-h committed:
feat(domain): 질문 및 답변 삭제 메소드 생성

Question
- Question에 delete(User user) 메소드 생성
  - 질문 삭제 권한 체크
  - 답변 삭제 권한 체크
  - 질문 삭제여부 true 변경
  - 답변 삭제여부 true 변경

earth-h committed:
feat(domain): 질문/답변 삭제이력 추가 메소드 생성

DeleteHistory
- 정적 팩토리 메소드 생성
  - LocalDateTime을 인자로 받는지에 따라 두가지 메소드로 나누어 사용
  - 기본적으로, 삭제이력 생성은 이력 남기는 그 순간 시간을 사용하여서 LocalDateTime 인자 없는 정적 팩토리 메소드를 비즈니스 로직에서 사용하도록 함
- 정적 팩토리 메소드 생성으로 인해, 기존에 사용하던 생성자는 private으로 변경

earth-h committed:
refactor(domain): CannotDeleteException을 extends RuntimeException으로 변경함에 따라 throws 부분 제거

earth-h committed:
refactor(domain,service): Question의 changeDeleted 메소드 private으로 변경 및 QnaService 리팩토링

Question
- changeDeleted() 메소드 결과 DeleteHistory 객체가 반환되어야 해서, 해당 메소드를 private으로 호출하도록 변경
  - 만약, changeDelete()를 public으로 열게되면, changeDelete()를 직접 호출하여서 삭제여부를 변경함에 따라 DeleteHistory 이력이 남지 않을 수 있기 때문

QuestionTest, QuestionRepositoryTest
- changeDelete() 메소드를 private으로 변경함에따라 테스트로직 수정

DeleteHistoryRepositoryTest
- Question의 delete() 메소드로 인해 DeleteHistory 객체 생성 테스트

Answer
- changeDeleted() 결과 DeleteHistory 반환하도록 변경

QnaService
- Question의 delete() 메소드 사용하도록 리팩토링

QnaServiceTest
- QnaService 내 deleteQuestion() 메소드 변경에 따른 mock 객체 수정

earth-h committed:
refactor(domain,service): Answer의 changeDeleted 메소드 private으로 변경

Answer
- changeDeleted() 메소드 결과 DeleteHistory 객체가 반환되어야 해서, 해당 메소드를 private으로 호출하도록 변경
  - 만약, changeDelete()를 public으로 열게되면, changeDelete()를 직접 호출하여서 삭제여부를 변경함에 따라 DeleteHistory 이력이 남지 않을 수 있기 때문
- changeDelete()을 private으로 만들면서, changeDelete()를 호출할 delete() 메소드 생성
  - 해당 메소드는 public으로 열려있으므로, validateSameUser()를 한 번 더 호출하여 혹 답변만 삭제하려는 상황 시, 작성자인지 체크 진행

Question
- Answer의 changeDelete() 메소드가 private이 됨에 따라 Answer의 delete() 호출하도록 변경

AnswerTest, AnswerRepositoryTest
- changeDelete() 메소드를 private으로 변경함에따라 테스트로직 수정

DeleteHistoryRepositoryTest
- Answer의 delete() 메소드로 인해 DeleteHistory 객체 생성 테스트 수정

earth-h committed:
feat(domain): 일급컬렉션을 가진 Answers 생성

Answers
- List<Answer> 리스트를 가짐
- Answer 삭제여부 변경 시, Question -> Answers -> Answer 순으로 메시지 보내도록 하고자 함

earth-h committed:
refactor(domain): Question의 addAnswer() 메소드 삭제

Question
- 답변 등록 시, toQuestion() 메소드 호출을 통해 특정 질문에 연결이 되므로, 질문 객체를 통해 답변 등록을 막음
  - 기본적으로 답변이 생길 때, 어떤 질문과 연결될지를 질문에서 추가되는게 맞다고 생각함

Answer
- toQuestion() 메소드가 외부에서 호출되지 않도록 private으로 변경

earth-h committed:
refactor(domain): Question에 List<Answer> 대신 Answers 사용하도록 리팩토링

Question
- for loop를 돌면서 질문에 대한 답변을 validation하고 답변의 삭제여부 변경하던 작업을 Answers가 대신하도록 수정

Answers
- Answers에 리스트 remove와 add 메소드 생성

Answer
- Question의 getter를 이용해서 remove/add하던 부분을 Question에 메시지를 보내서 할 수 있도록 변경

QnaServiceTest
- deleteHistory 리스트 생성 시, 답변 삭제이력을 먼저 생성하게 되어서, verify 테스트 순서 변경

earth-h committed:
refactor(domain): Question의 changeDeleted() 메소드를 deleted 필드값만 수정하는 역할을 갖도록 리팩토링

Question
- createDeleteHistory() 메소드를 생성하여, changeDeleted()가 가지고 있던 DeleteHistory 생성 역할을 넘김

earth-h committed:
refactor(domain): Answers에서 validateSameUser() 하는 부분 제거

Answers
- 답변 각각의 작성자를 검증하던 로직 제거
  - Answer에 메시지를 보내 삭제여부 변경을 할 때에 동일한 검증 작업 진행
    - 만약 답변 중 하나라도 검증에 실패하면 exception을 발생하여서 굳이 먼저 전체 답변에 대해 검증하지 않아도 되어 제거

DeleteHistoryRepositoryTest
- Answers에서 validation 구문 제거함에 따라 예외 제대로 발생하는지 테스트 추가

earth-h committed:
refactor(service): 불필요한 import 제거

earth-h committed:
refactor(domain): Answers에 존재하던 getter 제거

Answers
- getAnswers() 제거한 후 List<Answer>를 조회한 후에 하는 작업을 메소드로 각각 새로 생성
- List는 참조변수이기 때문에, 그냥 값을 꺼내게되면 add(), remove()가 가능할 수 있어 로직 내에서 반드시 필요한 이유가 없다면 추가하지 않는게 좋을 것으로 판단함

AnswerTest
- answers.getAnswers().size()로 호출하던 부분을 answers에 메시지를 보내 size()를 가져오도록 answers.answersCount() 메소드 추가하여 적용

Question
- question.getAnswers().size()로 호출하던 부분을 question.answersCount()로 변경하여서 answers에 메시지를 보내서 답을 얻어오도록 수정

earth-h committed:
test(service): 답변 없는 질문 delete 테스트 추가

QnaServiceTest
- 답변 없는 질문 delete 테스트 추가

earth-h committed:
feat(domain): DeleteHistories 일급 컬랙션 추가 및 Question의 delete() 메소드 내 순서 변경

DeleteHistories
- List<DeleteHistory>를 관리하는 일급 컬렉션
- 질문 삭제이력과 답변 삭제이력들을 더하는 merge() 메소드 생성
- DeleteHistoryRepository에 의해 저장할 용도로 deleteHistories를 꺼낼 때 수정할 수 없도록, unmodifiableList()로 꺼냄
- DeleteHistories 객체 생성 시, 수정할 수 없도록 복사생성자 사용

Question
- delete() 시, 질문 먼저 삭제여부 변경하고 답변 삭제여부 변경하고자 순서 변경

earth-h committed:
fix(domain): 답변 validation으로 인해 삭제여부 변경 못할 경우 질문 삭제여부 변경되면 안되므로 question의 delete() 메소드 동작 순서 재 변경 및 그에 따른 DeleteHistories 메소드 추가

Question
- 질문/답변 모두 validation 검증 이후 삭제여부 변경하도록 delete() 순서 변경

Answers
- validateSameUser()를 answer에서 delete할 때만 수행하고자 했으나, 그럴 경우 답변1,2가 있을 때 답변1은 validate 통과해서 답변1의 삭제여부가 참이 되는 현상 발생
  - 관련 테스트: QnaServiceTest의 "질문자와_답변자_달라_예외_발생하면_질문_삭제여부는_거짓"

DeleteHistories
- question의 delete() 메소드 내 순서 변경에 따라 기존에 merge() 메소드 대신 add() 메소드 생성

AnswerRepositoryTest, DeleteHistoriesTest, QnaServiceTest
- question의 delete() 메소드 동작 순서 관련 테스트
  - 질문/답변 validation 모두 통과해야 각각의 삭제여부 변경하는지 체크

earth-h committed:
refactor(service): deleteQuestion() 메소드 리팩토링

QnaService
- question.delete()의 반환값을 바로 deleteHistoryService.saveAll()의 인자값으로 사용하도록 리팩토링

earth-h committed:
feat(domain): Answer 생성 시, 삭제여부가 true인 질문에 생성된 답변인지 체크 로직 추가

Answer
- 삭제된 질문에 답변할 수 없으므로, Answer 생성 시, 질문 삭제여부 먼저 체크

ErrorCode
- NotFoundException 발생 시, 메시지도 함께 보일 수 있도록 에러코드 추가

NotFoundException
- 메시지도 함께 보일 수 있도록 생성자 추가

QnaService
- NotFoundException 생성자 변경에 따른 수정

earth-h committed:
refactor(domain): Answers 생성자 리팩토링

Answers
- Answers 객체 생성 시, list를 복사하여 넣도록 수정

earth-h committed:
refactor(domain): Answer 생성자 리팩토링

Answer
- 생성자 내 메소드 뽑을 수 있는 부분 추가로 뽑음
  - Question 관련 validation 메소드 생성

mskangg commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -81,8 +89,20 @@ public boolean isDeleted() {
+
+    private DeleteHistory changeDeleted(boolean deleted) {
         this.deleted = deleted;
+        return DeleteHistory.createDeleteHistory(ContentType.ANSWER, this.id, this.writer);
```
삭제 히스토리를 각각 `DeleteHistory.ofAnswer(id, writer) 와 ofQuestion` 으로 네이밍해서 만든다면 더욱 좋을 것 같아요 😃

mskangg commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public DeleteHistories delete(User user) {
+        List<DeleteHistory> deleteHistories = new ArrayList<>();
+        validateSameUser(user);
+        for(Answer answer: this.answers) {
+            deleteHistories.add(answer.delete(user));
+        }
+        return new DeleteHistories(deleteHistories);
```
```java
    public DeleteHistories deleteAll(User user) {
        return answers.stream()
                .map(answer -> answer.delete(user))
                .collect(Collectors.collectingAndThen(Collectors.toList(), DeleteHistories::new));
    }
```
stream을 활용해보시는건 어떠신가요??

earth-h replied:
오..! collectingAndThen()은 한 번도 안 써봤는데 감사합니다! 👍🏻 

answers를 stream()으로 만든 뒤, map()을 통해 delete()를 연산하고 그 결과를 다시 한 변 변환할 수 있게 하는 역할을 하는 거 같네요..!
그 후에 DeleteHistories를 생성할 수 있군요!!

감사합니다 ㅎㅎ 덕분에 새로운거 알게되었습니다 :)

mskangg commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,54 @@
+        return new DeleteHistories(deleteHistories);
+    }
+
+    private void validateSameUser(User user) {
+        for(Answer answer: this.answers) {
+            answer.validateSameUser(user);
+        }
+    }
```
> 상황) 질문A에 대한 답변1, 답변2가 존재할 때 답변2의 작성자가 질문과 상이하다.
이때, answer 내에서만 validateSameUser() 체크 시, 답변1은 validation 통과 후 삭제여부가 true가 되고, 이후 답변2는 validation에서 걸려 exception이 터진다. 
=> 답변1의 삭제여부가 다시 false로 변경되지 않는 문제 발생

이 부분은 서비스레이어에서 `@Transactional` 로 관리되어 있기 때문에 all pass / all fail 을 보장합니다.
그렇기 때문에 위 상황의 경우 답변1의 삭제여부가 다시 false 로 롤백되어집니다.

earth-h replied:
@Transactional에 의해서 롤백되어질 것을 기대하고 하기와 같이 테스트코드를 작성해서 진행했었는데요!
제가 원하는 결과가 나오지 않는 것 같아 validation을 중복으로 처리하였습니다 ㅠ
혹시 테스트할 때에 @Tranactional을 통해 롤백되었는지까지 테스트할 방법이 있을까요..? 

**[ QnaServiceTest.java ]**
```java
    @Test
    void 질문자와_답변자_달라_예외_발생하면_질문_삭제여부는_거짓() {
        //given
        User writer = TestUserFactory.create("sanjigi");
        User fakeWriter = TestUserFactory.create("javajigi");
        Question question = TestQuestionFactory.create(writer);
        Answer answer1 = TestAnswerFactory.create(writer, question);
        Answer answer2 = TestAnswerFactory.create(fakeWriter, question);
        questionRepository.save(question);

        //when
        when(questionRepository.findByIdAndDeletedFalse(question.getId())).thenReturn(Optional.of(question));
        assertThat(question.isDeleted()).isFalse();
        assertThat(question.answersCount()).isEqualTo(2);
        assertThatThrownBy(() -> qnaService.deleteQuestion(writer, question.getId()))
                .isInstanceOf(CannotDeleteException.class);

        //then
        assertThat(question.isDeleted()).isFalse();
        assertThat(answer1.isDeleted()).isFalse();
    }
```

**[ Answers.java ]**
```java
    public DeleteHistories delete(User user) {
        List<DeleteHistory> deleteHistories = new ArrayList<>();
        validateSameUser(user); // 이 부분을 주석처리하면 위 테스트코드 성공하지 않음
        for(Answer answer: this.answers) {
            deleteHistories.add(answer.delete(user));
        }
        return new DeleteHistories(deleteHistories);
    }

mskangg replied:
`@Transactional`은 도메인의 영역이 아닙니다.
이 부분은 QnaService에서 테스트하시면 될 것 같아요!

earth-h replied:
넵, 테스트는 QnaServiceTest에서 진행했습니다..! 

질문자와_답변자_달라_예외_발생하면_질문_삭제여부는_거짓()
요 테스트를 통해서 진행해보고자 한 것인데, assertThatThrownBy()를 통해 에러 발생하는 것은 확인하였는데,
그 이후 롤백여부가 assertThat()을 통해서 isFalse()에서 느껴지길 바랐는데 그게 느껴지지 않아서요..!

mskangg replied:
> [ QnaServiceTest.java ]

여기서 테스트했다고 작성하셨군요. 못봤네요 😅
그럼에도 발생안하는 이유는 @Transactional 자체가 spring aop로 동작하는데, 현재 모킹하여 테스트하기 때문에 동작하지 않는게 당연합니다.
해당 부분이 테스트가 필요하다면 @SpringBootTest 로 빈을 구동하면 동작할 것입니다.

earth-h replied:
오.. 넵넵! 확인해서 테스트해봤습니다 ㅠㅠ 모킹하면 @Transactional이 동작하지 않는 다는 사실을 인지하지 못했네요!! 
감사합니다 !!!

mskangg commented on src/main/java/qna/domain/DeleteHistories.java:
```diff
@@ -0,0 +1,40 @@
+        this.deleteHistories.add(deleteHistory);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        DeleteHistories that = (DeleteHistories) o;
+        return Objects.equals(deleteHistories, that.deleteHistories);
+    }
```
equals를 오버라이딩하신 이유가 있으신가요??
컬렉션이기 때문에 element 의 순서가 바뀐다면 다른 컬렉션으로 인식합니다.

earth-h replied:
기본적으로 제공되는 테스트코드와 제가 구현한 테스트코드 일부에서 DeleteHistories를 비교해야 하는 부분이 있어 구현하였습니다.

말씀해주신 컬렉션 순서에 따른 equals() 참/거짓이 달라질 수 있는 부분은 제가 고려하지 못했던 부분이라 해당 부분 추가하여서 equals() 오버라이딩 하도록 하겠습니다.

테스트케이스에서 체크할 때, DeleteHistories 내의 리스트를 뽑아서 체크할 수도 있지만, 그렇게되면 DeleteHistoryService의 saveAll() 인자로 리스트를 넘겨야합니다. 인자로 리스트를 감싼 일급 컬렉션을 넘기는게 더 명확하지 않을까 생각하여 그렇게 구현하였습니다..!

```java
    public void saveAll(DeleteHistories deleteHistories) {
        deleteHistoryRepository.saveAll(deleteHistories.unmodifiedDeleteHistories());
    }
```

감사합니다 :)

mskangg reviewed (changes requested):
3단계 미션도 잘 수행해주셨습니다 👍 
미션을 수행하면서 항상 중점적인 부분들 PR 로 설명해주셔서 보기간편했어요 🤩
몇가지 피드백 드렸으니 적용 부탁드려요 🙇‍♂️

> Answers에 생성한, removeAnswer()와 addAnswer()가 question에서 호출하기 때문에 public으로 열려있는데, 이렇게 열어도 될지 고민입니다. 외부 요인에 의해 잘못된 add나 remove를 막으려면 private이어야 할 것 같은데, 이를 내부적으로 수행할 방안을 찾지 못하였습니다..😭

사실 코드레벨에서 `외부 요인에 의해 잘못된 add나 remove`를 완전히 막을 순 없습니다.
add나 remove의 전제조건이 있다면 validation 정도 처리하는 것으로도 충분해 보여요 👍 

> Answer 각각의 validation 체크를 Answers에서도 진행한 부분

이 부분은 피드백 드렸습니당 😁

그리고 요구사항 중 하나인 `모든 원시 값과 문자열을 포장한다` 를 만족하지 못하는것 같습니다!
contents, name, email 등... 많이 포장할 수 있을 것 같네요 🤔

mskangg commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -56,4 +58,44 @@ public class AnswerTest {
+        //when
+        assertThatThrownBy(() -> TestAnswerFactory.create(writer, question)).isInstanceOf(NotFoundException.class)
+                .hasMessage(ErrorCode.삭제된_질문에는_답변할_수_없음.getErrorMessage());
+    }
```
질문이_존재하지_않음 에 대한 예외 테스트케이스도 추가해보아요 😃

earth-h committed:
refactor(domain): DeleteHistory 정적 팩토리 메소드 네이밍 변경(피드백 반영)

DeleteHistory
- 기존에 createDeleteHistory()로 질문과 답변 삭제이력 모두에 대해 같이 쓰던 메소드를 질문용, 답변용으로 분리해 ContentType을 따로 인자로 받지 않게됨

earth-h committed:
refactor(domain): DeleteHistories equals() 메소드 수정

DeleteHistories
- equals() 내 리스트 순서 영향 받지 않도 동일한 원소들로 구성되었는지 확인하기 위해 equals() 메소드 수정

DeleteHistoriesTest
- DeleteHistories의 equals()가 리스트원소 순서 달라도 정상 작동하는지 확인하기 위함

earth-h committed:
test(domain): validation 테스트케이스 추가

AnswerTest
- 존재하지 않는 질문에 대한 답변 생성 시 발생할 수 있는 예외 테스트케이스 추가

earth-h committed:
refactor(domain): Answers의 loop돌면서 delete() 호출하고 DeleteHistories 생성하던 부분 stream 적용(피드백 적용)

Answers
- stream.map()을 통해 먼저 delete()을 수행하고, 해당 결과값을 collectionAndThen을 통해 먼저 list로 만든 후 DeleteHistories 생성자를 통해 신규 DeleteHistories 객체 생성

earth-h committed:
feat(domain): 원시값 포장(Contents)

Contents
- Contents는 String을 포장하였고, Contents 내의 contents 내용이 같으면 같은 Contents로 봐야하므로 equals() 오버라이딩
  - String contents는 같은 내용이면 같은 객체를 반환할 것이기 때문
- Answer, Question의 toString이 있으므로, 두 객체를 toString으로 부를 때 Contents도 제대로 toString() 하고자 오버라이딩

earth-h committed:
test(domain): 원시값 포장(Contents) 테스트 추가

ContentsTest
- 테스트코드 추가 및 테스트코드 보강

earth-h committed:
feat(domain): 원시값 포장(Title)

Title
- Title은 String을 포장하였고, String은 같은 내용이면 동일한 객체를 반환할 것이므로, equals() 오버라이딩
- toString() 오버라이딩

earth-h committed:
feat(domain): 원시값 포장(Name)

Name
- Name은 String을 포장하였고, String은 같은 내용이면 동일한 객체를 반환할 것이므로, equals() 오버라이딩
- toString() 오버라이딩

earth-h committed:
feat(domain): 원시값 포장(Password)

Password
- Password은 String을 포장하였고, String은 같은 내용이면 동일한 객체를 반환할 것이므로, equals() 오버라이딩
- toString() 오버라이딩

earth-h committed:
feat(domain): 원시값 포장(Email)

Email
- Email은 String을 포장하였고, String은 같은 내용이면 동일한 객체를 반환할 것이므로, equals() 오버라이딩
- toString() 오버라이딩

earth-h committed:
feat(domain): 원시값 포장(UserId), User 내에 존재하던 Password와 UserId 관련 validation을 각 객체에 메시지 보내어 진행하도록 수정

UserId
- UserId은 String을 포장하였고, String은 같은 내용이면 동일한 객체를 반환할 것이므로, equals() 오버라이딩
- toString() 오버라이딩
- User에 있던 UserId 관련 validation을 UserId에 메시지 보내도록 수정

Password
- User에 있던 Password 관련 validation을 UserId에 메시지 보내도록 수정

User
- 기존에 User 내에서 메소드로 진행하던 UserId, Password 관련 validation 수정
- UserId 적용

earth-h committed:
fix(domain): User 내 userId, password validation 부분 잘못 매핑된 파라미터 수정

earth-h committed:
refactor(domain): User내 name과 email 관련 동일성 체크 하는 부분을 각각 객체에 메시지 보내어 하도록 수정

User
- isEqualName(), isEqualEmail() 부분 제거 및 객체로 메시지 보내도록 수정

Email
- isEqualEmail() 추가

Name
- isEqualName() 추가

earth-h committed:
refactor(test): 불필요한 import 제거 및 assertThat() 여러개인 테스트를 assertAll()로 묶음

earth-h committed:
refactor(domain): for loop를 forEach로 변경

earth-h committed:
refactor(domain): validationSameUser() 중복 제거

@SpringBootTest를 통해 rollback 확인하여 중복 제거

earth-h commented:
안녕하세요, 민수님!

주신 피드백 바탕으로 수정하여서 재 피드백 요청드립니다~!

> 그리고 요구사항 중 하나인 모든 원시 값과 문자열을 포장한다 를 만족하지 못하는것 같습니다!
contents, name, email 등... 많이 포장할 수 있을 것 같네요 🤔

✏︎ 원시값과 문자열인데 원시값만이라고 보고 넘어가버렸네요 ㅠㅠ 이 부분 추가하였습니다 !

> 삭제 히스토리를 각각 DeleteHistory.ofAnswer(id, writer) 와 ofQuestion 으로 네이밍해서 만든다면 더욱 좋을 것 같아요 😃

✏︎ 각각 질문용/답변용으로 생성하여서 넘겨야할 파라미터도 줄고, 명확한 이름을 통해 DeleteHistory를 만들 수 있었습니다.

> stream을 활용해보시는건 어떠신가요??

✏︎ 덕분에 collectingAndThen()을 알게되었습니다 ! 감사합니다 👍🏻 

> 이 부분은 서비스레이어에서 @Transactional 로 관리되어 있기 때문에 all pass / all fail 을 보장합니다.
그렇기 때문에 위 상황의 경우 답변1의 삭제여부가 다시 false 로 롤백되어집니다.

✏︎ mock 하는 부분을 단순하게 특정 행위 결과를 강제해주는 느낌으로 인지하고 있었는데, 모킹으로 인해 @Transactional이 동작하지 않는 다는 점은 처음 알게 되었습니다..! Spring AOP와 mock을 얘기해주셔서, 두 키워드로 검색을 해봤는데 아직 정확한 정보는 찾지 못했네용.. ㅎㅎ.. 이 부분은 좀 더 공부해보겠습니다 :) 

> equals를 오버라이딩하신 이유가 있으신가요??
컬렉션이기 때문에 element 의 순서가 바뀐다면 다른 컬렉션으로 인식합니다.

✏︎ DeleteHistories에 equals()를 오버라이딩한 이유는, DeleteHistories의 saveAll() 메소드의 인자로 List<DeleteHistory>가 아닌 랩핑된 일급컬렉션 객체를 넘기기 위함이었습니다 ! 순서가 바뀔 수 있는 부분에 대해서는, equals()에서 비교할 때의 로직을 변경해두었습니다 ! 피드백 감사합니다 👍🏻 

추가적으로, 문자열 포장 및 말씀해주신 테스트코드 추가 필요한 부분에 대해서는 추가하였습니다 !

감사합니다 :)

earth-h committed:
refactor(domain): 문자열 래핑 시, 길이 제한 validation 추가

mskangg reviewed (approved):
모든 요구사항을 만족시킨 100점 짜리 코드군요~! 💯 
깐깐한 리뷰 받아주셔서 감사합니다 🙇‍♂️
우테캠은 `일이 아닌 학습이기에` 자칫 오버엔지니어링이 될 수도 있는 경험을 해보는 것도 이 과정에서 느낄 수 있을거라 `모든 원시 값과 문자열을 포장한다`가 요구사항에 포함되어 있다고 생각합니다.
앞으로 미션 진행함에 있어서 `일이 아닌 학습으로` 생각하고 임해주신다면, 많은 것을 얻을 수 있으실거에요!

그럼 이만 머지해드리고 다음 미션 화이팅입니다! 🚀
